rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,388 @@
+        # The Buy asset and amount is stored in the last transfer log
+        buy_asset = self.base.get_or_create_evm_token(transfer_log.address)
+        buy_amount = int.from_bytes(transfer_log.data)
+        fee_asset = out_event.asset
+
+        # Remove the transfer event
+        # of the purchased token to the user's wallet
+        # to create a new event with the right sequence index
+        remove_action = ActionItem(
+            action='skip',
+            from_event_type=HistoryEventType.RECEIVE,
+            from_event_subtype=HistoryEventSubType.NONE,
+            asset=buy_asset,
+        )
+
+        tx_hash = transaction.tx_hash
+        timestamp = transaction.timestamp
+        amount = asset_normalized_value(buy_amount, self._resolve_asset(buy_asset))
+        in_event = self.base.make_event_next_index(
+            tx_hash=tx_hash,
+            timestamp=timestamp,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.RECEIVE,
+            asset=buy_asset,
+            amount=amount,
+            location_label=sender,
+            notes=f'Receive {amount} {buy_asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap',  # noqa: E501
+            counterparty=CPT_RAINBOW_SWAPS,
+            address=self.router_address,
+        )
+        decoded_events.append(in_event)
+
+        # Create fee event as the last event
+        fee_event = self._create_fee_event(
+            tx_hash=tx_hash,
+            sequence_index=in_event.sequence_index + 1,
+            timestamp=timestamp,
+            fee_raw=fee_raw,
+            fee_asset=fee_asset,
+            sender=sender,
+        )
+
+        return DecodingOutput(event=fee_event, action_items=[remove_action])
+
+    def _maybe_decode_swap_token_to_native(
+        self,
+        transaction: EvmTransaction,
+        all_logs: list[EvmTxReceiptLog],
+        in_event: Optional['EvmEvent'],
+        out_event: Optional['EvmEvent'],
+        sender: ChecksumEvmAddress,
+
+    ) -> DecodingOutput:
+        """
+        Decodes a swap transaction from token to native token.
+        Check if the first log is transfer and the last log is withdrawn
+        """
+        native_token = self.evm_inquirer.native_token
+        unwrap_log = all_logs[-1]
+        if (
+            (all_logs[0] != ERC20_OR_ERC721_TRANSFER and bytes_to_address(all_logs[0].topics[2]) == sender) or  # noqa: E501
+            unwrap_log.topics[0] != WETH_WITHDRAW_TOPIC
+        ):
+            return DEFAULT_DECODING_OUTPUT
+
+        if not (in_event and out_event):
+            return DEFAULT_DECODING_OUTPUT
+
+        try:
+            in_amount = asset_raw_value(in_event.amount, native_token)
+        except UnsupportedAsset:
+            in_amount = 0
+        # The fee will be the withdraw data subtract the in_amount
+        fee_raw = int.from_bytes(unwrap_log.data) - in_amount
```

Comment:
you would need to find a reliable way to find the unwrap log event. I believe you can check for transfers to the router contract as those are the initiators of the logic